p2p/libp2p/discovery/optimizedKadDhtDiscoverer.go

```diff
@@ -0,0 +1,231 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+)
+
+type discovererStatus string
+
+const statNotInitialized discovererStatus = "not initialized"
+const statInitialized discovererStatus = "initialized"
+const minIntervalForSeedersReconnection = time.Second
+const optimizedKadDhtName = "optimized kad-dht discovery"
+
+type optimizedKadDhtDiscoverer struct {
+	kadDHT                      KadDhtHandler
+	peersRefreshInterval        time.Duration
+	seedersReconnectionInterval time.Duration
+	protocolID                  string
+	initialPeersList            []string
+	bucketSize                  uint32
+	routingTableRefresh         time.Duration
+	hostConnManagement          *hostWithConnectionManagement
+	sharder                     Sharder
+	status                      discovererStatus
+	chanInit                    chan struct{}
+	errChanInit                 chan error
+	chanConnectToSeeders        chan struct{}
+	chanDoneConnectToSeeders    chan struct{}
+	initKadDhtHandler           func(ctx context.Context) (KadDhtHandler, error)
+}
+
+// NewOptimizedKadDhtDiscoverer creates an optimized kad-dht discovery type implementation
+// initialPeersList can be nil or empty, no initial connection will be attempted, a warning message will appear
+// This implementation uses a single process loop function able to carry multiple tasks synchronously
+func NewOptimizedKadDhtDiscoverer(arg ArgKadDht) (*optimizedKadDhtDiscoverer, error) {
+	sharder, err := prepareArguments(arg)
+	if err != nil {
+		return nil, err
+	}
+
+	if arg.SeedersReconnectionInterval < minIntervalForSeedersReconnection {
+		return nil, p2p.ErrInvalidSeedersReconnectionInterval
+	}
+
+	okdd := &optimizedKadDhtDiscoverer{
+		sharder:                     sharder,
+		peersRefreshInterval:        arg.PeersRefreshInterval,
+		seedersReconnectionInterval: arg.SeedersReconnectionInterval,
+		protocolID:                  arg.ProtocolID,
+		initialPeersList:            arg.InitialPeersList,
+		bucketSize:                  arg.BucketSize,
+		routingTableRefresh:         arg.RoutingTableRefresh,
+		status:                      statNotInitialized,
+		chanInit:                    make(chan struct{}),
+		errChanInit:                 make(chan error),
+		chanConnectToSeeders:        make(chan struct{}),
+		chanDoneConnectToSeeders:    make(chan struct{}),
+	}
+
+	okdd.initKadDhtHandler = okdd.initKadDht
+	okdd.hostConnManagement, err = NewHostWithConnectionManagement(arg.Host, okdd.sharder)
+	if err != nil {
+		return nil, err
+	}
+
+	go okdd.processLoop(arg.Context)
+
+	return okdd, nil
+}
+
+// Bootstrap will start the bootstrapping new peers process
+func (okdd *optimizedKadDhtDiscoverer) Bootstrap() error {
```

Comment:
Is there any logic flow that prevents this implementation in having the same function calls as `continuousKadDhtDiscoverer`? In the first implementation `Bootstrap` calls `startDHT` and will actually start the process loop, whereas in the new implementation the process loop is started in the constructor, `Bootstrap` only gives the go signal.

Having the same flow and will make it easier for anyone looking at the difference between the two implementations in parallel.

Reply:
The `continuousKadDhtDiscoverer` will be pending for removal (after this new implementation will be tested thoroughly) because it is harder to maintain and there are a lot of pitfalls there. 
Regarding the starting of the go routine I have decided to implement it in this way as to avoid mutexes (we need some sort of mutex usage and flags when starting the go routine as to not start it twice).